src/components/fields/number-field/number-field.spec.js

```diff
@@ -1,172 +1,210 @@
-  onChange: () => jest.fn(),
-  ...customProps,
-});
-
-describe('rendering', () => {
-  describe('data attributes', () => {
-    let textInput;
-    beforeEach(() => {
-      const props = createTestProps({
-        'data-foo': 'bar',
-        'data-test': 'baz',
-      });
-      const wrapper = shallow(<NumberField {...props} />);
-      textInput = wrapper.find(NumberInput);
-    });
-    it('should forward the attributes to the NumberInput', () => {
-      expect(textInput).toHaveProp('data-foo', 'bar');
-      expect(textInput).toHaveProp('data-test', 'baz');
-    });
+
+// This component is used to enable easy testing.
+// It overwrites the onChange function and places a label for the
+// input component. It also ensures an id so that the label can associate
+// the input. This allows tests to use getByLabelText.
+// It also makes sure the event's value passed to onChange flows back to the
+// component so that we can test it under real conditions.
+// As a convenience, we enable accessing a mocked onChange function.
+class Story extends React.Component {
+  static displayName = 'Story';
+  static propTypes = {
+    onEvent: PropTypes.func.isRequired,
+    value: PropTypes.string,
+    id: PropTypes.string,
+  };
+  static defaultProps = {
+    id: 'text-field',
+  };
+  state = {
+    value: this.props.value || '',
+  };
+  handleChange = event => {
+    this.props.onEvent(event);
+    this.setState({ value: event.target.value });
+  };
+  render() {
+    return (
+      <div>
+        <label htmlFor={this.props.id}>NumberField</label>
+        <NumberField
+          id={this.props.id}
+          {...omit(this.props, 'onEvent')}
+          value={this.state.value}
+          onChange={this.handleChange}
+        />
+      </div>
+    );
+  }
+}
+
+const renderNumberField = (customProps, options) => {
+  const onChange = jest.fn();
+  const props = {
+    title: 'foo',
+    ...customProps,
+  };
+  return {
+    onChange,
+    ...render(<Story onEvent={onChange} {...props} />, options),
+  };
+};
+
+it('should render a text field', () => {
+  const { getByLabelText } = renderNumberField();
+  expect(getByLabelText('NumberField')).toBeInTheDocument();
+});
+
+it('shoult render a title', () => {
+  const { getByText } = renderNumberField({ title: 'foo title' });
+  expect(getByText('foo title')).toBeInTheDocument();
+});
+
+it('should forward data-attributes', () => {
+  const { container } = renderNumberField({ 'data-foo': 'bar' });
+  expect(container.querySelector('[data-foo="bar"]')).toBeInTheDocument();
+});
+
+it('should have an HTML name', () => {
+  const { container } = renderNumberField({ name: 'foo' });
+  expect(container.querySelector('[name="foo"]')).toBeInTheDocument();
+});
+
+it('should call onFocus when the input is focused', () => {
+  const onFocus = jest.fn();
+  const { container } = renderNumberField({ onFocus });
+  container.querySelector('input').focus();
+  expect(container.querySelector('input')).toHaveFocus();
```

Comment:
The `Story` component renders a `label` which we can use to find the input instead of using `container.querySelector('input')`. This is closer to how a user would interact with the component and should therefore be preferred.

The *[Which get method should I use?](https://github.com/kentcdodds/dom-testing-library#faq)* section explains this nicely:

![image](https://user-images.githubusercontent.com/1765075/50145444-e8389500-02b1-11e9-97f8-75b3e138861c.png)